core/log/profiler_hook.cpp

```diff
@@ -372,12 +382,50 @@ std::shared_ptr<ProfilerHook> ProfilerHook::create_for_executor(
 }
 
 
+std::shared_ptr<ProfilerHook> ProfilerHook::create_summary(std::ostream& stream,
+                                                           std::string name,
+                                                           bool nested)
+{
+    auto fns = create_summary_fns(stream, name, nested);
+    return std::shared_ptr<ProfilerHook>{
+        new ProfilerHook{std::move(fns.first), std::move(fns.second)}};
+}
+
+
 std::shared_ptr<ProfilerHook> ProfilerHook::create_custom(hook_function begin,
                                                           hook_function end)
 {
     return std::shared_ptr<ProfilerHook>{new ProfilerHook{begin, end}};
 }
 
 
+/**
+ * Scope guard that annotates its scope with the provided profiler hooks.
+ */
+profiling_scope_guard::profiling_scope_guard(const char* name,
+                                             profile_event_category category,
+                                             ProfilerHook::hook_function begin,
+                                             ProfilerHook::hook_function end)
+    : empty_{false}, name_{name}, category_{category}, end_{end}
+{
+    begin(name, category);
+}
+
+profiling_scope_guard::~profiling_scope_guard()
+{
+    if (!empty_) {
+        end_(name_, category_);
+    }
+}
+
+profiling_scope_guard::profiling_scope_guard(profiling_scope_guard&& other)
```

Comment:
As long as we still use C++14, we need it to be able to return from a function (`ProfilerHook::user_range`), but with C++17, it will no longer be necessary